### ISSUE TITLE
unnecessary_to_owned causes type mismatch

### ISSUE THREAD

**Issue by Starwort** [rust-lang/rust-clippy#8929]
### Summary

Clippy will suggest turning `.to_vec().into_iter()` into `.iter().copied()` even when doing so causes a type mismatch later in the function

### Lint Name

unnecessary_to_owned

### Reproducer

I tried this code:

[full code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c69de5dd76445484f6df942e958006bb)
```rust
let mut foos_left = FOOS.to_vec().into_iter().filter(make_filter(true, 'b'));
foos_left = foos_left
    .collect::<Vec<_>>()
    .into_iter()
    .filter(make_filter(false, 'z'));
```

I saw this happen:

```
warning: unnecessary use of `to_vec`
  --> src/main.rs:17:25
   |
17 |     let mut foos_left = FOOS.to_vec().into_iter().filter(make_filter(true, 'b'));
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `FOOS.iter().copied()`
   |
   = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
```

I expected to see this happen:
Nothing; applying the change causes a type mismatch error because the type of `foos_left` changes

### Version

```text
rustc 1.63.0-nightly (e09449220 2022-05-31)
binary: rustc
commit-hash: e0944922007e1bb4fe59809293acf4364410cccc
commit-date: 2022-05-31
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.4
```

### Additional Labels

@rustbot label +l-suggestion-causes-error

**Comment by Starwort** [comment]
> **Error**: Label l-suggestion-causes-error can only be set by Rust team members

(why is this mentioned in the form field then?)

**Comment by hellow554** [comment]
@Starwort it's an uppercase i, not a lowercase L (:

@rustbot modify labels: I-suggestion-causes-error

**Comment by Starwort** [comment]
> @Starwort it's an uppercase i, not a lowercase L (:

Ah, of course

**Comment by smoelius** [comment]
Not long ago, we changed the suggestion to `MaybeIncorrect` because of [a similar problem](https://github.com/rust-lang/rust-clippy/issues/8148): https://github.com/rust-lang/rust-clippy/blob/0d5ace3ed22fdd6ee25ee1dda42139d36c0aeec4/clippy_lints/src/methods/unnecessary_to_owned.rs#L221-L223

**Comment by H2CO3** [comment]
I experienced the same bug while linting [the following type and function](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6eaec14dc2256266f240f6c7ec77fc68):

```rust
enum FewOrMany {
    Few([u16; 4]),
    Many(Box<[u16]>),
}

impl IntoIterator for FewOrMany {
    type IntoIter = <Vec<u16> as IntoIterator>::IntoIter;
    type Item = u16;

    fn into_iter(self) -> Self::IntoIter {
        match self {
            FewOrMany::Few(buf) => buf.to_vec().into_iter(),
            FewOrMany::Many(buf) => Box::into_iter(buf),
        }
    }
}
```

**Comment by shssoichiro** [comment]
I'm also experiencing a similar error with the following, where `remaining_sites` is a `HashSet<String>`: `remaining_sites.contains(&"python_docs".to_string())`

**Comment by H2CO3** [comment]
> I'm also experiencing a similar error with the following, where `remaining_sites` is a `HashSet<String>`: `remaining_sites.contains(&"python_docs".to_string())`

That doesn't seem like a false positive, though. Since `String: Borrow<str>`, you can use a string literal (and a slice in general) to look up items in a `HashSet<String>` or a `HashMap<String, _>`: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=79856c52aa171f2fdf91f32da23f97ea)

**Comment by shssoichiro** [comment]
Ah, I found out why using a `str` wasn't working in my case. It's actually a `HashSet<&String>` which doesn't seem to have the same Borrow-friendly behavior. (Which, I should fix the type of the HashSet, although it is technically a case where the lint's suggestion leads to failed compilation without the extra step of changing the HashSet type.)
